src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs

```diff
@@ -17,7 +20,19 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     internal partial class DiagnosticAnalyzerService : IIncrementalAnalyzerProvider
     {
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
-            => _map.GetValue(workspace, _createIncrementalAnalyzer);
+        {
+            if (GlobalOptions.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode))
+            {
+                // We rely on LSP to query us for diagnostics when things have changed and poll us for changes that might
+                // have happened to the project or closed files outside of VS.
+                // However, we still need to create the analyzer so that the map contains the analyzer to run when pull diagnostics asks.
+                _ = _map.GetValue(workspace, _createIncrementalAnalyzer);
+
+                return NoOpIncrementalAnalyzer.Instance;
```

Comment:
Before we get into these options, I wanted to quickly check with you if there is anything in the LSP client side that would be caching diagnostics. One of the core advantages of our diagnostic incremental analyzer service is that it caches all the diagnostics, and repeated queries will hit the cache and avoid recomputation. I think we want to preserve the caching behavior somewhere - either in the clients OR a shared intermediate component, if any, that is pulling the diagnostics OR retain the solution crawler based diagnostic incremental analyzer to compute diagnostics in the background, but disable raising any push events for diagnostics changed from it. This way we turn off push model but still have the background analysis computing and caching diagnostics when appropriate, and the LSP pull diagnostics gets the caching benefits. I don't know if that sounds like a reasonable behavior to you.

Getting back to your suggested options, I agree that if the goal is to determine just whether or not active document diagnostics have changed, listening to workspace changed events sounds reasonable. Any degradation can only come for non-active documents, as editing a method body in active document would not affect diagnostics for non-active documents (for majority of analyzers/compiler diagnostics at least), but that doesn't seem to be our concern when determining when to tell the editor to re-query us for new set of suggestion actions.

Reply:
While we don't directly cache the last diagnostic set, we do know when something has changed and we only ask the diagnostic service for diagnostics when it has.  See https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs,140

Basically we look at the solution checksums to see if we might need to recompute diagnostics.  This means for repeated requests (as happens when the client polls) as long as nothing has changed we're not doing any kind of recomputation.  Is that essentially the caching you are looking for, or is it more involved in the incremental analyzer?

> R retain the solution crawler based diagnostic incremental analyzer to compute diagnostics in the background, but disable raising any push events for diagnostics changed from it. This way we turn off push model but still have the background analysis computing and caching diagnostics when appropriate, and the LSP pull diagnostics gets the caching benefits. I don't know if that sounds like a reasonable behavior to you

this is similar to what happens now, but I think leads us to potentially computing diagnostics twice.  Basically when the workspace moves forward, the sln crawler will ask to recompute diagnostics at the same time LSP is polling us.  But they may be on slightly different solutions each request.  The second thing that makes me not want to go this route is that Tomas' plan is to eventually remove the solution crawler, at which point we'll have to figure out a different caching mechanism anyway.  So if the current caching I'm doing isn't good enough, I'd rather improve that.

> Any degradation can only come for non-active documents, as editing a method body in active document would not affect diagnostics for non-active documents (for majority of analyzers/compiler diagnostics at least), but that doesn't seem to be our concern when determining when to tell the editor to re-query us for new set of suggestion actions.

Sounds good!  Yeah in that scenario I think the act of opening the non-active file should trigger a suggested actions query anyway, so hopefully it still works.  I'll test things out though and make sure.  Thanks!

Reply:
> Basically we look at the solution checksums to see if we might need to recompute diagnostics

Sounds good.

> Is that essentially the caching you are looking for, or is it more involved in the incremental analyzer?

I think that should be sufficient and matches what we do compute the version in diagnostic analyzer service [here](https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs,208 ), which is used to fetch the cached diagnostics [here](https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs,25)

> The second thing that makes me not want to go this route is that Tomas' plan is to eventually remove the solution crawler

This would be awesome!